Fix c10 registration binary size

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26865 [wip] Remove manual boxing wrappers
* #26882 Make TypeDefault, TypeDerived and VariableType anonymous namespaces
* #26878 Fix binary size in schema inference
* #26869 Fix shared_ptr binary size in op registration
* #26860 Improve binary size of function schema inference
* **#26827 Fix c10 registration binary size**

The templates there had a binary size impact of ~25MB. This PR fixes that.

Differential Revision: [D17566642](https://our.internmc.facebook.com/intern/diff/D17566642/)

smessmer committed:
[wip] Fix FunctionSchemaInferer binary size



Differential Revision: [D17566642](https://our.internmc.facebook.com/intern/diff/D17566642/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Fix FunctionSchemaInferer binary size"


Differential Revision: [D17566642](https://our.internmc.facebook.com/intern/diff/D17566642/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Fix FunctionSchemaInferer binary size"


Differential Revision: [D17566642](https://our.internmc.facebook.com/intern/diff/D17566642/)

[ghstack-poisoned]

resistor commented on aten/src/ATen/core/op_registration/infer_schema.h:
```diff
@@ -94,7 +94,7 @@ FunctionSchema createFunctionSchemaFromTraits(std::string name, std::string over
 template<class FuncType>
-FunctionSchema inferFunctionSchema(std::string name, std::string overload_name) {
+FunctionSchema inferFunctionSchema(std::string&& name, std::string&& overload_name) {
   return detail::createFunctionSchemaFromTraits<guts::infer_function_traits_t<FuncType>>(std::move(name), std::move(overload_name));
```
Are these moves actually doing anything?  Can't we pass this through as a reference all the way to the innermost frame?

smessmer replied:
That's why I changed it to `&&`, now it's just passed through as a reference.

smessmer replied:
update: changing the `FunctionSchema` constructor reduces by ~100-200kb but that seems within regular build size fluctuation.

smessmer replied:
I implemented it in https://github.com/pytorch/pytorch/pull/26860, maybe it helps

smessmer committed:
Update on "Fix c10 registration binary size"


The templates there had a binary size impact of ~25MB. This PR fixes that.

Differential Revision: [D17566642](https://our.internmc.facebook.com/intern/diff/D17566642/)

[ghstack-poisoned]

ezyang commented:
Tentatively removing myself as reviewer

smessmer committed:
Update on "Fix c10 registration binary size"


The templates there had a binary size impact of ~25MB. This PR fixes that.

Differential Revision: [D17566642](https://our.internmc.facebook.com/intern/diff/D17566642/)

[ghstack-poisoned]

resistor reviewed (approved):
Verified the code size benefit
